### ISSUE TITLE
FR: Option to autostart Tailscale after AndroidTV system restart

### ISSUE THREAD

**Issue by Heib386** [tailscale/tailscale#7824]
### What are you trying to do?

Autstart Tailscale after Android TV reboots

### How should we solve this?

Option to autostart tailscale

### What is the impact of not solving this?

no workaround atm

### Anything else?

would be really helpful to get this feature

**Comment by kjvalencik** [comment]
It would be great to have an autostart option on android devices that don't have access to Always-on VPN. For example, firestick.

I did some quick hacking and it _should_ be possible.

* Changed `autoConnect = false` to `autoConnect = true` in `App.java`. This causes the app to automatically start on launch.
* Installed [Autostart - No Root](https://play.google.com/store/apps/details?id=com.autostart&hl=en_US&gl=US) and configured Tailscale to start on boot

This was sufficient to get Tailscale connecting on boot.

The Autostart app appears to be using the `android.intent.action.BOOT_COMPLETED`. The Tailscale could also listen on this intent (as well as adding the permission). The current IPNReceiver requires the app to be opened at least once because it doesn't initialize the Go modules or networks.

**Comment by apexiptv** [comment]
@kjvalencik I am wondering this myself. I have googleTV's in all my airbnb's.. Sometimes I need to remote in and update app logins or troubleshoot things when people staying have issues with it. I use vysor adb to do this with the help of Tailscale but I have no way to get the Tailscale app to autostart or auto-connect when it's opened.

**Comment by ellnic** [comment]
I would really appreciate this also. I use Nvidia shield's and the auto start feature would be great. The iOS app has recently had an on demand setting added, which works really well. The android TV app could use a bit of love next ;)

@kjvalencik can you link to your apk? :)

**Comment by sbkg0002** [comment]
Any news on this? Any simple workaround?

**Comment by strarsis** [comment]
@kjvalencik: I noticed your workaround for Firestick, thanks! However, on the Firestick I noticed that it first shows some recommended ads immediately after boot/restart. Only after dismissing the initial ad the app autostarts. Is there a workaround for this? Are the initial ads also caused by an app, and can the app somehow have a higher priority on startup and start before the ads?

**Comment by kjvalencik** [comment]
I haven't noticed that issue. I have auto play and personalized ads disabled.

Which device do you have? I have a 4k max.

**Comment by strarsis** [comment]
@kjvalencik: 
> I have auto play and personalized ads disabled.

I disabled both options in recommended content section.


> Which device do you have? I have a 4k max.

The device in my case is a `FireTV Stick Lite`.

**Comment by sbkg0002** [comment]
The same applies for the Nokia Streaming Box 8000.

**Comment by stephenjharrington** [comment]
Any time line for this?
It's really annoying that it doesn't auto-start after a reboot, in my case an nVidia Shield TV Pro.

**Comment by sisland** [comment]
FYI: Using the latest tailscale available on Firestick (1.66) works perfectly. I just removed and plugged in several times and always keep the latest state and exit node too. So, in my case: The enabled + Exit node works after a cold/hard reboot.

Also, using same apk version 1.66 I sideloaded on an Android TV (Sony Bravia 2020) and it works as on the firestick

**Comment by stewnet** [comment]
> FYI: Using the latest tailscale available on Firestick (1.66) works perfectly. I just removed and plugged in several times and always keep the latest state and exit node too. So, in my case: The enabled + Exit node works after a cold/hard reboot.
> 
> Also, using same apk version 1.66 I sideloaded on an Android TV (Sony Bravia 2020) and it works as on the firestick

This kind of works for a google tv with no option in android settings for always-on VPN. Unfortunately it won't auto connect on boot, you still have to open the app, but you don't have to scroll up the the slider to connect once in the app.
Would really love an auto connect option. I saw a comment from the tailscale devs a few weeks ago that the tailscale app should be in the play store for TVos devices soon, so hopefully they address this issue with that release.

**Comment by baltazartroisville** [comment]
The FireTV now has an official version in the Amazon Appstore. It works perfectly, a small cosmetic bug is that there are big black borders on the left and right.
The only annoying thing that is still not fixed is the autostart feature, at the moment you have to open the app by hand for it to connect automatically.
Apps like Adguard connect automatically after rebooting FireTV, but I have no idea how they do it.

**Comment by Pawloland** [comment]
This can be remedied by using adb. More info here https://github.com/tailscale/tailscale/issues/510#issuecomment-2220310581

**Comment by baltazartroisville** [comment]
From the last issue, here again summarized and explained for everyone.

tl;dr:

To enable the Always-On VPN feature for Tailscale:
```shell
adb shell settings put secure always_on_vpn_app com.tailscale.ipn
```

Explanation:

With an ADB command it is possible to enable the Always-On VPN feature of Android for Tailscale. This feature should be available since Android 7 or FireOS 6.

The following steps are required to enable it:

1. install ADB (Android Debug Bridge).

2. change to the ADB directory using the Windows CMD or the Windows or Linux Terminal using the CD command.

3. find out the IP address of your Android TV or FireTV. Enable ADB debugging in the TV device settings.

4. run the following command and replace "IP" with the local IP of your device (for Windows, add ".exe" after "adb") and then allow the connection on the TV:
```shell
adb connect IP
```

5. run the following command
```shell
adb shell settings put secure always_on_vpn_app com.tailscale.ipn
```

The issue mentions two other commands, but as far as I know they are not necessary because they are already disabled by default, but for the sake of completeness I list them here. The first disables the VPN lockdown feature, i.e. no data can be transferred without an active VPN connection. If this were active (set to 1) and no exit node was set up in Tailscale, Internet access would no longer work, which is logical since Tailscale does not normally route normal Internet connections outside its own tailnet. The second command removes all applications that are exempt from the VPN lockdown feature.
```shell
adb shell settings put secure always_on_vpn_lockdown 0
```
```shell
adb shell settings put secure always_on_vpn_lockdown_whitelist ''
```

Have a nice day everyone! ❤️

**Comment by Immanuel31** [comment]
It seems that for Android TV the [Always-on VPN](https://support.google.com/android/answer/9089766#zippy=%2Cstay-connected-all-the-time) option requires adb "hackery" and is not available through a regular Android TV Settings menu.  Could the option be added in the Tailscale Settings menu?  Consider this a FR / Feature Request :)

**Comment by tsquillario** [comment]
I'd like to send a Chromecast to someone (my dad) configured and ready to go. As soon as he plugs it in and connects to WiFi, he should be on my Tailnet. Seems like there is plenty of demand for this feature and it makes sense put it the Settings UI

**Comment by yamanobe96** [comment]
It's been a year since this request was made, but has there been any progress?

Engineers are happy to take the time to easily turn tailscale on,

but many of the people who actually use it after setup are elderly parents and children who have no software knowledge.

It's cruel to ask them to determine whether tailscale is on or off and turn it on manually.

**Comment by bobmarley2021** [comment]
@yamanobe96 Whilst I too have been eagerly awaiting for this feature (and a couple of others) I think you have to understand that catering for the children and elderly relatives of people [most probably on the free plan] is not going to be a number one priority. I don't think you can say Tailscale are being 'cruel', such a comment is juvenile, so I assume you mean you are being cruel to them? Why not choose a different solution in the meantime if you are concerned about how you're treating your family?

With that said, you are correct when you said that this has been open for a long time, but if you'd looked at the year it's nearly 2 years in fact.

**Comment by mygrexit** [comment]
> From the last issue, here again summarized and explained for everyone after successful testing.

Thank you so much for that!
Do you guys have any experience if always_on_vpn_app will prevent the tv from going into sleep/deep sleep?

**Comment by Mick2011** [comment]
+1 for wanting this to autostart at boot.

**Comment by Immanuel31** [comment]
The [adb command above](https://github.com/tailscale/tailscale/issues/7824#issuecomment-2416123677) works great for a 'Chromecast with Google TV' to the point that I had forgotten about it.
Planning an upgrade I have now bought a '[Google TV Streamer (4K)](https://store.google.com/product/google_tv_streamer)' and I successfully submitted the very same ..
```
adb shell settings put secure always_on_vpn_app com.tailscale.ipn
``` 
.. and Tailscale seems to start automatically, but it does not succeed in connecting to my Tailnet: It keeps spinning :(
Please note that Tailscale has no problem connecting when starting it manually on the new 'Google TV Streamer (4K)'.

All is still working fine for the 'Chromecast with Google TV' running the Android TV OS build UTTC.250917.004.14193937.
While it does not work automatically for the 'Google TV Streamer (4K)' running the Android TV OS build UTTK.250729.004.

**UPDATE:** Upgrading the Tailscale app seems to have fixed the issue. It is now starting and connecting automagically!
[Google Play](https://play.google.com/store/apps/details?id=com.tailscale.ipn) offered an upgrade from Version 1.90.4-t0d7298602-g92f2bb273 to Version 1.90.9-t6e8a4f2de-g2dcab7d21

**Comment by ghost** [comment]
For those who don't want to do this with adb, a simple method is by using a custom launcher like projectivity. Click on app > start on boot.

**Comment by pippo73** [comment]
hello all,
I've experienced the same problem for a bug that seems it has more then 3 years. Is there any chance to fix it?
